### PR TITLE
CSHARP-3033: Support speculative MONGODB-X509 authentication attempts in isMaster

### PR DESCRIPTION
**Pull request by vincentkam**
~~https://evergreen.mongodb.com/version/5ea2faf27742ae3a3e7e170d~~
https://evergreen.mongodb.com/version/5eac70449ccd4e024fd6a2fe
https://jira.mongodb.org/browse/CSHARP-3033

### CODE REVIEW

File: src/MongoDB.Driver.Core/Core/Authentication/MongoDBX509Authenticator.cs [hunk 1]
```diff
@@ -69,6 +69,11 @@ public void Authenticate(IConnection connection, ConnectionDescription descripti
             Ensure.IsNotNull(description, nameof(description));
             EnsureUsernameIsNotNullOrNullIsSupported(connection, description);
 
+            if (description.IsMasterResult.SpeculativeAuthenticate != null)
+            {
+                return;
```
Comment (vincentkam) [inline src/MongoDB.Driver.Core/Core/Authentication/MongoDBX509Authenticator.cs:74]: Java also simply returns when it sees that the `speculativeAuthenticate` subdocument exists. https://github.com/jstewart-mongo/mongo-java-driver/pull/112/files#diff-7612d2ebaf5b02d03270c4731591fbefR45

File: tests/MongoDB.Driver.Tests/AuthenticationTests.cs [hunk 2]
```diff
@@ -132,7 +134,7 @@ public void Authentication_succeeds_when_user_has_Scram_Sha_1_Mechanism_and_mech
             settings.Credential = MongoCredential
                 .FromComponents(mechanism: null, source: null, username: userName, password: password);
 
-            AssertAuthenticationSucceeds(settings);
+            AssertAuthenticationSucceeds(settings, shouldSpeculativelyAuthenticateIfPossible: false);
```
Comment (vincentkam) [inline tests/MongoDB.Driver.Tests/AuthenticationTests.cs]: We speculatively authenticate with scramsha256 when the authmechanism isn't specified
https://github.com/mongodb/specifications/blob/master/source/mongodb-handshake/handshake.rst#speculative-authentication, but we only  have a scramsha1 user and we did NOT specify the mechanism, so we expect  `isMasterResult` to NOT have a `speculativeAuthenticate` subdocument.

**Comment by vincentkam** [comment]
Ready for review!

File: tests/MongoDB.Driver.Tests/AuthenticationTests.cs [hunk 3]
```diff
@@ -347,6 +352,16 @@ private void AssertAuthenticationSucceeds(MongoClientSettings settings, bool asy
                 {
                     client.ListDatabaseNames().ToEnumerable().ToList();
                 }
+                if (shouldSpeculativelyAuthenticateIfPossible &&
```
Comment (vincentkam) [inline tests/MongoDB.Driver.Tests/AuthenticationTests.cs]: Note: this technically adds a speculative auth test to all the other tests in this file except for the `Authentication_succeeds_when_user_has_Scram_Sha_1_Mechanism_and_mech`

File: src/MongoDB.Driver.Core/Core/Authentication/MongoDBX509Authenticator.cs (hunk 1 above)
Comment (vincentkam) [reply src/MongoDB.Driver.Core/Core/Authentication/MongoDBX509Authenticator.cs:74]: See also: https://github.com/mongodb/specifications/blob/master/source/auth/auth.rst#mongodb-x509

File: tests/MongoDB.Driver.Tests/AuthenticationTests.cs [hunk 4]
```diff
@@ -347,6 +352,16 @@ private void AssertAuthenticationSucceeds(MongoClientSettings settings, bool asy
...
+                    CoreTestConfiguration.ServerVersion >= Feature.SpeculativeAuthentication.FirstSupportedVersion)
```
Comment (DmitryLukyanov) [inline tests/MongoDB.Driver.Tests/AuthenticationTests.cs]: Should it be `Feature.SpeculativeAuthentication.IsSupported`?

File: tests/MongoDB.Driver.Tests/AuthenticationTests.cs [hunk 5]
```diff
@@ -24,6 +24,8 @@
 using MongoDB.Bson;
 using MongoDB.Bson.Serialization.Serializers;
 using MongoDB.Bson.TestHelpers.XunitExtensions;
+using MongoDB.Driver.Core.Clusters.ServerSelectors;
+using MongoDB.Driver.Core.Connections;
```
Comment (DmitryLukyanov) [inline tests/MongoDB.Driver.Tests/AuthenticationTests.cs]: not used ns. Also, please remove all other not used `ns`s here

**Review by DmitryLukyanov — APPROVED** [review]
Minor comments(including the question from java chat) and then LGTM

File: tests/MongoDB.Driver.Tests/AuthenticationTests.cs [hunk 6]
```diff
@@ -347,6 +352,16 @@ private void AssertAuthenticationSucceeds(MongoClientSettings settings, bool asy
                 {
                     client.ListDatabaseNames().ToEnumerable().ToList();
                 }
+                if (speculativeAuthenticatationShouldSucceedIfPossible &&
```
Comment (vincentkam) [inline tests/MongoDB.Driver.Tests/AuthenticationTests.cs]: Moving outdated comment: Note: this technically adds a speculative auth test to all the other tests in this file except for the Authentication_succeeds_when_user_has_Scram_Sha_1_Mechanism_and_mech

File: tests/MongoDB.Driver.Tests/AuthenticationTests.cs [hunk 7]
```diff
@@ -132,7 +134,7 @@ public void Authentication_succeeds_when_user_has_Scram_Sha_1_Mechanism_and_mech
             settings.Credential = MongoCredential
                 .FromComponents(mechanism: null, source: null, username: userName, password: password);
 
-            AssertAuthenticationSucceeds(settings);
+            AssertAuthenticationSucceeds(settings, speculativeAuthenticatationShouldSucceedIfPossible: false);
```
Comment (vincentkam) [inline tests/MongoDB.Driver.Tests/AuthenticationTests.cs]: Moving outdated comment: We speculatively authenticate with scramsha256 when the authmechanism isn't specified
https://github.com/mongodb/specifications/blob/master/source/mongodb-handshake/handshake.rst#speculative-authentication, but we only have a scramsha1 user and we did NOT specify the mechanism, so we expect isMasterResult to NOT have a speculativeAuthenticate subdocument.

File: tests/MongoDB.Driver.Tests/AuthenticationTests.cs (hunk 5 above)
Comment (vincentkam) [reply tests/MongoDB.Driver.Tests/AuthenticationTests.cs]: Done

File: tests/MongoDB.Driver.Tests/AuthenticationTests.cs (hunk 4 above)
Comment (vincentkam) [reply tests/MongoDB.Driver.Tests/AuthenticationTests.cs]: Done

File: tests/MongoDB.Driver.Tests/AuthenticationTests.cs (hunk 5 above)
Comment (vincentkam) [reply tests/MongoDB.Driver.Tests/AuthenticationTests.cs]: Done

**Comment by rstam** [comment]
I'm waiting for you to finish

https://github.com/vincentkam/mongo-csharp-driver/pull/106

and to rebase this pull request on that one before reviewing this one.

**Review by rstam — CHANGES_REQUESTED** [review]
Please re-request my review when https://github.com/vincentkam/mongo-csharp-driver/pull/106 is done and pushed to master.

**Comment by vincentkam** [comment]
Rebased and re-requesting review!

File: src/MongoDB.Driver.Core/Core/Authentication/MongoDBX509Authenticator.cs (hunk 1 above)
Comment (vincentkam) [reply src/MongoDB.Driver.Core/Core/Authentication/MongoDBX509Authenticator.cs:74]: Verified offline with John and Dima in the java channel that simply returning here should be correct

File: tests/MongoDB.Driver.Tests/AuthenticationTests.cs [hunk 8]
```diff
@@ -453,9 +462,10 @@ private void DropDatabaseUser(MongoClient client, string database, string userNa
         private string GetRfc2253FormattedUsernameFromX509ClientCertificate(X509Certificate2 certificate)
         {
             var distinguishedName = certificate.SubjectName.Name;
-            // Authentication will fail if we don't remove the spaces, even if we add the username WITH the spaces.
-            var nameWithNoSpaces = string.Join(",", distinguishedName.Split(',').Select(s => s.Trim()));
-            return nameWithNoSpaces.Replace("S=", "ST=");
+            // Authentication will fail if we don't remove the delimiting spaces, even if we add the username WITH the
+            // delimiting spaces.
+            var nameWithoutDelimitingSpaces = string.Join(",", distinguishedName.Split(',').Select(s => s.Trim()));
```
Comment (vincentkam) [inline tests/MongoDB.Driver.Tests/AuthenticationTests.cs:468]: A slightly more descriptive name and comments

File: src/MongoDB.Driver.Core/Core/Authentication/MongoDBX509Authenticator.cs (hunk 1 above)
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Authentication/MongoDBX509Authenticator.cs:74]: I don't understand why we don't have to examine the SpeculativeAuthenticate value for errors.

If authentication fails due to invalid credentials where is that turned into an authentication exception?

File: src/MongoDB.Driver.Core/Core/Authentication/MongoDBX509Authenticator.cs [hunk 9]
```diff
@@ -101,18 +111,22 @@ public async Task AuthenticateAsync(IConnection connection, ConnectionDescriptio
...
         }
 
         // private methods
+
+        private BsonDocument CreateAuthenticateCommand() => new BsonDocument
+        {
+            { "authenticate", 1 },
+            { "mechanism", Name },
+            { "user", _username, _username != null }
+        };
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Authentication/MongoDBX509Authenticator.cs]: In general I would prefer expression bodied methods to be limited to one-liners.

Specially bad in this case is that the way the opening and closing braces are formatted makes this "look" like a traditional method.

If you really prefer an expression bodied method even for something that can't fit in one line at least format it to make it visually obvious that this method is an expression bodied method and not a traditional method:

    private BsonDocument CreateAuthenticateCommand()
        =>
            new BsonDocument
            {
                { "authenticate", 1 },
                { "mechanism", Name },
                { "user", _username, _username != null }
            };

**Review by rstam — CHANGES_REQUESTED** [review]
Just a couple questions/comments.

**Comment by vincentkam** [comment]
Quick note: I think we can ignore the failures in EG for sharded clusters: we've filed a bug for it: https://jira.mongodb.org/browse/SERVER-47908

File: src/MongoDB.Driver.Core/Core/Authentication/MongoDBX509Authenticator.cs (hunk 1 above)
Comment (vincentkam) [reply src/MongoDB.Driver.Core/Core/Authentication/MongoDBX509Authenticator.cs:74]: Consider the non-speculative logic below: 
https://github.com/vincentkam/mongo-csharp-driver/pull/110/files#diff-1df9e524f12ce43cd79dedf55798c65dR102
Here, we create the first command in the conversation described [here](https://github.com/mongodb/specifications/blob/master/source/auth/auth.rst#mongodb-x509)

The server's reply will either contain `ok: 0` or `ok: 1`. In the case of `ok: 0`, a `MongoCommandException` is thrown and we catch and convert the exception  to a `MongoAuthenticationException` on line 107, calling `CreateException`.

The process is similar for speculative authentication, except that there is no `ok` field in the `speculativelyAuthenticate` subdocument in the isMaster reply as noted [here](https://github.com/mongodb/specifications/blob/master/source/mongodb-handshake/handshake.rst#speculative-authentication):

> The speculativeAuthenticate reply has the same fields, except for the ok field, as seen in the conversation sections for MONGODB-X509, SCRAM-SHA-1 and SCRAM-SHA-256 in the Driver Authentication spec.

As with non speculative X509 auth,  a successful X509 speculative auth conversation only involves a single client message and a single server message.  Thus, in the case of X509 speculative auth, the very presence of the `speculativeAuthenticate` subdocument in the isMaster reply should indicate we've successfully authenticated, unlike in SCRAM-SHA-X, where additional steps are needed to finish authentication.

Additionally, my understanding is that the bulk of the security of using x509 authentication comes from TLS itself. If the client certificate is invalid, I believe the handshake with the mongod will not even complete. All the driver is doing identifying what user is associated with said client certificate to the server, in the case of servers < 3.4. For severs >3.4, the driver doesn't even need to supply the username.

File: src/MongoDB.Driver.Core/Core/Authentication/MongoDBX509Authenticator.cs (hunk 9 above)
Comment (vincentkam) [reply src/MongoDB.Driver.Core/Core/Authentication/MongoDBX509Authenticator.cs]: I've reformatted this the "traditional" way :)

**Review by vincentkam — COMMENTED** [review]
Read for re-review!

File: src/MongoDB.Driver.Core/Core/Authentication/MongoDBX509Authenticator.cs (hunk 1 above)
Comment (vincentkam) [reply src/MongoDB.Driver.Core/Core/Authentication/MongoDBX509Authenticator.cs:74]: If the TLS handshake succeeds, but MONGODB-X509 auth fails (perhaps due to the user not existing in the `$external` database) then speculative authentication should fail, resulting in an isMaster response without `speculativeAuthenticate` document. We then fallback to normal authentication, which will result in a server reply with `ok: 0`, triggering a `MongoCommandException` which we will then catch (e.g. on line 107), converting that exception to a `MongoAUthenticationException`. If you'd like, we could add a test to assert that this exception is thrown as expected?

Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Authentication/MongoDBX509Authenticator.cs:74]: I see now... if the speculative authentication fails then the entire `isMaster` command returns `ok : 0`.

**Review by rstam — APPROVED** [review]
LGTM

File: src/MongoDB.Driver.Core/Core/Authentication/MongoDBX509Authenticator.cs (hunk 1 above)
Comment (vincentkam) [reply src/MongoDB.Driver.Core/Core/Authentication/MongoDBX509Authenticator.cs:74]: Discussing this offline!

**Comment by vincentkam** [comment]
Squashed, rebased, and merged to master.

**Pull request closed without merging**